fuzzgen: Mostly Forward Branching

afonso360 opened the pull request:
👋 Hey,

This PR addresses some of the issues found in #3050 regarding the performance of fuzzgen. We found that we had an outsized number of timeouts during execution of the fuzzer. This is mostly due to generating a bunch of infinite loops.

This changes our branching strategy. Instead of choosing blocks at random, we always choose a block ahead of the current block. This makes it so that we always make forward progress and don't time out very often. I still think backwards branches are important to generate, so we make those a configurable option (0.1% of branches).

Additionally it looks like not having valid target branches was also an issue. 27.3% of our invalid `Arbitrary::choice` failures were due to us trying to select a block that we didn't have. Such as for example trying to select a block that does not have params when all of them do.
We are now smarter about which terminators we allow to be inserted avoiding these issues.

cc: @jameysharp 

## Benchmarks

<details>
<summary><h3>Baseline</h3></summary>

Statistics:
```
#49967  NEW    cov: 31776 ft: 183224 corp: 4225/5444Kb lim: 394895 exec/s: 23 rss: 966Mb L: 2106/373965 MS: 2 ShuffleBytes-CopyPart-
== FuzzGen Statistics  ====================
Total Inputs: 50000
Valid Inputs: 18378 (36.8%)
Total Runs: 62769
Successful Runs: 48229 (76.8% of Total Runs)
Timed out Runs: 10440 (16.6% of Total Runs)
Traps:
        user code: int_divz: 4100 (6.5% of Total Runs)
        user code: bad_sig: 0 (0.0% of Total Runs)
        user debug: 0 (0.0% of Total Runs)
        user code: icall_null: 0 (0.0% of Total Runs)
        user code: heap_oob: 0 (0.0% of Total Runs)
        user code: int_ovf: 0 (0.0% of Total Runs)
        user code: bad_toint: 0 (0.0% of Total Runs)
        user code: interrupt: 0 (0.0% of Total Runs)
        user code: stk_ovf: 0 (0.0% of Total Runs)
        user code: table_oob: 0 (0.0% of Total Runs)
        user code: heap_misaligned: 0 (0.0% of Total Runs)
        user code: unreachable: 0 (0.0% of Total Runs)
        resumable: 0 (0.0% of Total Runs)

```

`Arbitrary::choice` failures:
```
Loaded 20139 stacks

First Line duplicates
5.9 - Count: 1181 - Line:    0: cranelift_fuzzgen::function_generator::FunctionGenerator::generate_br_table
8.3 - Count: 1679 - Line:    0: cranelift_fuzzgen::function_generator::FunctionGenerator::generate_jumptables
1.4 - Count: 282 - Line:    0: cranelift_fuzzgen::function_generator::FunctionGenerator::generate_switch
11.7 - Count: 2355 - Line:    0: cranelift_fuzzgen::function_generator::FunctionGenerator::generate_target_block
29.9 - Count: 6014 - Line:    0: cranelift_fuzzgen::function_generator::FunctionGenerator::get_variable_of_type
29.7 - Count: 5991 - Line:    0: cranelift_fuzzgen::function_generator::FunctionGenerator::stack_slot_with_size
13.1 - Count: 2637 - Line:    0: cranelift_fuzzgen::function_generator::insert_call
```

</details>

<details>
<summary><h3>This PR</h3></summary>

Statistics:
```
#49979  NEW    cov: 31503 ft: 178604 corp: 3520/7374Kb lim: 1048576 exec/s: 48 rss: 1054Mb L: 2702/1048576 MS: 1 CMP- DE: "\x01\x00\x00\x00\x00\x00\x00\x05"-
== FuzzGen Statistics  ====================
Total Inputs: 50000
Valid Inputs: 21036 (42.1%)
Total Runs: 724599
Successful Runs: 719276 (99.3% of Total Runs)
Timed out Runs: 32 (0.0% of Total Runs)
Traps:
        user code: stk_ovf: 0 (0.0% of Total Runs)
        user code: heap_misaligned: 0 (0.0% of Total Runs)
        user code: bad_sig: 0 (0.0% of Total Runs)
        user code: table_oob: 0 (0.0% of Total Runs)
        user code: bad_toint: 0 (0.0% of Total Runs)
        user code: heap_oob: 0 (0.0% of Total Runs)
        user debug: 0 (0.0% of Total Runs)
        resumable: 0 (0.0% of Total Runs)
        user code: icall_null: 0 (0.0% of Total Runs)
        user code: int_ovf: 0 (0.0% of Total Runs)
        user code: int_divz: 5291 (0.7% of Total Runs)
        user code: unreachable: 0 (0.0% of Total Runs)
        user code: interrupt: 0 (0.0% of Total Runs)
```

`Arbitrary::choice` failures:
```
Loaded 14367 stacks

First Line duplicates
41.3 - Count: 5936 - Line:    0: cranelift_fuzzgen::function_generator::FunctionGenerator::get_variable_of_type
37.6 - Count: 5395 - Line:    0: cranelift_fuzzgen::function_generator::FunctionGenerator::stack_slot_with_size
21.1 - Count: 3036 - Line:    0: cranelift_fuzzgen::function_generator::insert_call
```
</details>

This improves execs/s by 108% (23 -> 48) which is nice. It also eliminates all invalid choices regarding block terminators so we reject fewer inputs as well.

jameysharp commented on cranelift/fuzzgen/src/function_generator.rs:
```diff
@@ -736,8 +942,19 @@ where
-        let block_targets = &self.resources.blocks[1..];
+        // We try to mostly generate forward branches to avoid generating an excessive amount of
+        // infinite loops. But they are still important, so give them a 0.1% chance of existing.
+        let generate_backwards = self.u.ratio(1, 1000)?;
+        let (backwards_blocks, forward_blocks) = self.resources.partition_blocks(source_block);
+        let block_targets = if generate_backwards && !backwards_blocks.is_empty() {
```
It's a minor thing, but I'd prefer to not consume any fuzz input if there aren't any backwards blocks to jump to:

```suggestion
        let (backwards_blocks, forward_blocks) = self.resources.partition_blocks(source_block);
        let block_targets = if !backwards_blocks.is_empty() && self.u.ratio(1, 1000)? {
```

jameysharp commented on cranelift/fuzzgen/src/function_generator.rs:
```diff
@@ -542,6 +542,161 @@ const OPCODE_SIGNATURES: &'static [(
+
+    let default_block = {
+        let target_blocks = fgen.resources.forward_blocks_without_params(source_block);
+        *fgen.u.choose(target_blocks)?
```
Seems unfortunate that `insert_switch` can't generate backward branches, but I guess that's fine.

afonso360 replied:
Yeah, we should probably include that

jameysharp replied:
I don't mind leaving this as something to come back to later though. If you want to just add a TODO comment to `insert_switch` that it would be nice to also check backward branches, that's plenty.

jameysharp reviewed (commented):
This is looking like a huge improvement. I'm delighted that you can measure just how much of an improvement it is! I just have a few suggestions.

afonso360 commented on cranelift/fuzzgen/src/function_generator.rs:
```diff
@@ -571,6 +726,57 @@ struct Resources {
+
+    /// Generates a slice of `blocks_without_params` ahead of `block`
+    fn forward_blocks_without_params(&self, block: Block) -> &[Block] {
+        let partition_point = self
+            .blocks_without_params
+            .iter()
+            .enumerate()
+            .find(|(_, b)| **b > block)
+            .map_or(self.blocks_without_params.len(), |(i, _)| i);
+
+        let (_, forward_blocks) = self.blocks_without_params.split_at(partition_point);
+        forward_blocks
```
That is much neater, thanks! I wasn't aware of `partition_point()`

afonso360 commented on cranelift/fuzzgen/src/function_generator.rs:
```diff
@@ -571,6 +726,57 @@ struct Resources {
+        &self,
+        block: Block,
+    ) -> (&[(Block, BlockSignature)], &[(Block, BlockSignature)]) {
+        let partition_point = self
+            .blocks
+            .iter()
+            .enumerate()
+            .find(|(_, (b, _))| *b > block)
+            .map_or(self.blocks.len(), |(i, _)| i);
```
Yeah, that's a great suggestion! I was able to get this to the point where we directly index into the blocks vector using the block number. 

We don't necessarily need to store the block alongside the Signature, however removing it makes it so that we can't return a slice here, and the caller of this is quite messy without it and without allocating anything.

jameysharp replied:
Oh, good point! Storing the block number does make things easier in the caller.

afonso360 commented:
Hey, I think I addressed most of the feedback above (Thanks for the great review!).

I've also done some additional changes:
* Added the `backwards_block_ratio` to the config
* Moved all of the block generation code to inside `generate_blocks`

jameysharp commented on cranelift/fuzzgen/src/function_generator.rs:
```diff
@@ -593,6 +745,40 @@ struct Resources {
+            .iter()
+            .copied()
+            .filter(|jt| jump_tables[*jt].iter().all(|target| *target > block))
+            .collect()
```
On further thought, we could avoid allocating here by sorting `self.jump_tables` in `generate_jumptables`. I'm happy to merge without it but I'm going to write down the idea in case you decide you want to do it.

The sort key is `|jt| builder.func.jump_tables[*jt].iter().min()`. In `generate_jumptables` you could use that as the argument to `self.resources.jump_tables.sort_by_cached_key(...)`. Then here you could use it in `self.jump_tables.partition_point(... <= block)`.

To keep the sort key computation in only one place, you could have `generate_jumptables` save a pair of `(min_block, table_id)` in `self.resources.jump_tables`. Then the sort is just `self.resources.jump_tables.sort_unstable_by_key(|(min_block, _)| *min_block)` and this function just needs `self.jump_tables.partition_point(|(min_block, _)| *min_block <= block)`.

afonso360 replied:
That sounds quite doable! I'm going to add a TODO comment to revisit later, I'd prefer to tackle the other fuzzgen issues for now.

jameysharp reviewed (approved):
This is fantastic. Thank you! I'm looking forward to seeing how this affects fuzz performance once we merge it.

Let me know when you've fuzzed to your satisfaction and I'll be happy to merge.

afonso360 commented:
Something went wrong in the rebase and its now crashing, I'm going to take a look at this again tomorrow.

afonso360 commented:
So, it turns out nothing went wrong in the rebase and it was a preexisting issue with this PR. We were sometimes selecting `block0` as a target for branching, which makes the verifier very unhappy. I didn't notice this because these inputs were silently being rejected and #4896 finally exposed this.

I think this may explain the difference between the results in this PR and the original stats in https://github.com/bytecodealliance/wasmtime/issues/3050#issuecomment-1243037213. Although here we have 10k stacks where we don't know what happened and that seems quite a bit.

I'm re running the benchmarks to see if this discrepancy has gone away.

afonso360 commented:
So, here are the results.

This is using a custom patch on top of the arbitrary crate to report failures. See https://github.com/bytecodealliance/wasmtime/issues/3050#issuecomment-1244761070 for more details.

<details>
<summary><h3>Baseline</h3></summary>

```
#99826  NEW    cov: 32083 ft: 189603 corp: 5384/10156Kb lim: 1048576 exec/s: 10 rss: 1030Mb L: 2586/1048576 MS: 4 EraseBytes-ChangeASCIIInt-ShuffleBytes-PersAutoDict- DE: "Q\x03\x00\x00\x00\x00\x00\x0
0"-
== FuzzGen Statistics  ====================
Total Inputs: 100000
Valid Inputs: 61534 (61.5%)
Total Runs: 426680
Successful Runs: 211464 (49.6% of Total Runs)
Timed out Runs: 34331 (8.0% of Total Runs)
Traps:
        user code: int_divz: 180882 (42.4% of Total Runs)
        user code: heap_misaligned: 0 (0.0% of Total Runs)
        user code: icall_null: 0 (0.0% of Total Runs)
        user code: bad_toint: 0 (0.0% of Total Runs)
        resumable: 0 (0.0% of Total Runs)
        user code: int_ovf: 3 (0.0% of Total Runs)
        user code: interrupt: 0 (0.0% of Total Runs)
        user debug: 0 (0.0% of Total Runs)
        user code: stk_ovf: 0 (0.0% of Total Runs)
        user code: bad_sig: 0 (0.0% of Total Runs)
        user code: unreachable: 0 (0.0% of Total Runs)
        user code: heap_oob: 0 (0.0% of Total Runs)
        user code: table_oob: 0 (0.0% of Total Runs)
```

```
[BUILD] stack.rs
Loaded 10000
Loaded 20000
Loaded 30000
Loaded 38472 stacks  
                     
First Line duplicates
100.0 - Count: 38472 - Line: EmptyChoose

Two Line duplicates
6.3 - Count: 2415 - Line: EmptyChoose
   0: arbitrary::trace_error
   1: arbitrary::unstructured::Unstructured::choose_index
   2: cranelift_fuzzgen::function_generator::FunctionGenerator::generate_br_table
8.3 - Count: 3199 - Line: EmptyChoose
   0: arbitrary::trace_error
   1: arbitrary::unstructured::Unstructured::choose_index
   2: cranelift_fuzzgen::function_generator::FunctionGenerator::generate_jumptables
1.4 - Count: 534 - Line: EmptyChoose
   0: arbitrary::trace_error
   1: arbitrary::unstructured::Unstructured::choose_index
   2: cranelift_fuzzgen::function_generator::FunctionGenerator::generate_switch
20.2 - Count: 7779 - Line: EmptyChoose
   0: arbitrary::trace_error
   1: arbitrary::unstructured::Unstructured::choose_index
   2: cranelift_fuzzgen::function_generator::FunctionGenerator::generate_target_block
35.2 - Count: 13526 - Line: EmptyChoose
   0: arbitrary::trace_error
   1: arbitrary::unstructured::Unstructured::choose_index
   2: cranelift_fuzzgen::function_generator::FunctionGenerator::get_variable_of_type
21.9 - Count: 8413 - Line: EmptyChoose
   0: arbitrary::trace_error
   1: arbitrary::unstructured::Unstructured::choose_index
   2: cranelift_fuzzgen::function_generator::FunctionGenerator::stack_slot_with_size
6.8 - Count: 2606 - Line: EmptyChoose
   0: arbitrary::trace_error
   1: arbitrary::unstructured::Unstructured::choose_index
   2: cranelift_fuzzgen::function_generator::insert_call
```
</details>


<details>
<summary><h3>This PR</h3></summary>

```
#99990  REDUCE cov: 31890 ft: 186525 corp: 4925/10508Kb lim: 393599 exec/s: 19 rss: 1034Mb L: 144/316625 MS: 1 EraseBytes-
== FuzzGen Statistics  ====================
Total Inputs: 100000
Valid Inputs: 74409 (74.4%)
Total Runs: 1540097
Successful Runs: 1240133 (80.5% of Total Runs)
Timed out Runs: 356 (0.0% of Total Runs)
Traps:
        user code: unreachable: 0 (0.0% of Total Runs)
        user code: stk_ovf: 0 (0.0% of Total Runs)
        user code: int_divz: 299534 (19.4% of Total Runs)
        user code: heap_oob: 0 (0.0% of Total Runs)
        user code: bad_toint: 0 (0.0% of Total Runs)
        user code: int_ovf: 74 (0.0% of Total Runs)
        user code: table_oob: 0 (0.0% of Total Runs)
        resumable: 0 (0.0% of Total Runs)
        user debug: 0 (0.0% of Total Runs)
        user code: heap_misaligned: 0 (0.0% of Total Runs)
        user code: icall_null: 0 (0.0% of Total Runs)
        user code: bad_sig: 0 (0.0% of Total Runs)
        user code: interrupt: 0 (0.0% of Total Runs)
```

```
[BUILD] stack.rs
Loaded 10000
Loaded 20000
Loaded 25591 stacks

First Line duplicates
100.0 - Count: 25591 - Line: EmptyChoose

Two Line duplicates
59.0 - Count: 15098 - Line: EmptyChoose
   0: arbitrary::trace_error
   1: arbitrary::unstructured::Unstructured::choose_index
   2: cranelift_fuzzgen::function_generator::FunctionGenerator::get_variable_of_type
29.9 - Count: 7645 - Line: EmptyChoose
   0: arbitrary::trace_error
   1: arbitrary::unstructured::Unstructured::choose_index
   2: cranelift_fuzzgen::function_generator::FunctionGenerator::stack_slot_with_size
11.1 - Count: 2848 - Line: EmptyChoose
   0: arbitrary::trace_error
   1: arbitrary::unstructured::Unstructured::choose_index
   2: cranelift_fuzzgen::function_generator::insert_call
```

</details>

1. Wow! #4895 makes a big difference in `int_divz` traps!

2. For baseline we have `38472` stacks and `(100000-61534)=38466` invalid inputs. That is close enough that I'm going to call it valid. I suspect the 6 extra stacks come from the fact that I run this with `-runs=100010` and 6 out of those 10 extra executions failed.

I think this should be good for merging now. We no longer see the discrepancy in stacks.

jameysharp commented:
This is fantastic. To summarize your statistics:
- 21% increase in valid inputs
- more than 3x as many total runs while doubling executions per second
- almost 6x as many successful runs
- eliminated almost 99% of timeouts
